File: backend/graphs/util/proportions.py
import pandas as pd
pd.DataFrame.iteritems = pd.DataFrame.items
# These will let us use R packages:
from rpy2.robjects import pandas2ri
from rpy2.robjects.vectors import StrVector
from rpy2.robjects import conversion
from rpy2.robjects.packages import STAP
# Database interaction
from sqlalchemy import Engine, MetaData
from util.sql_queries import basic_selection

pandas2ri.activate()

def proportions(engine: Engine, meta: MetaData, table_name: str, column: str | None, values: list[str], word_list: list[str]):
    # Import proportions function from proportions.R
    with open("graphs/util/proportions.R", "r") as file:
        proportions = file.read()
    proportions = STAP(proportions, "proportions")

    data = basic_selection(engine, meta, table_name, column, values, word_list)
    # Run proportions function
    if column != None and len(values) > 0:
        # Goup by word and group (if defined)
        # Store ids as list
        ids = data.groupby(["word", "group"])["id"].apply(list).reset_index(name = "ids")
        data.drop("id", axis = 1, inplace = True)
        output = data.groupby(["word", "group"]).sum().reset_index()
        group_counts = output.groupby("group")["count"].sum().reset_index()
        group_counts.rename(columns = { "count": "total" }, inplace = True)
        output = pd.merge(output, group_counts, on = "group")
        output["proportion"] = output["count"] / output["total"]
        output.drop(["count", "total"], axis = 1, inplace = True) 
            
        # Add ids
        output["ids"] = ids["ids"]
    else:
        # If groups not defined
        ids = data.groupby(["word"])["id"].apply(list).reset_index(name = "ids")
        data.drop("id", axis = 1, inplace = True)
        output = data.groupby()
        output["proportion"] = output["count"] / sum(output["count"])
        output.drop("count", axis = 1, inplace = True) 
    # output = conversion.rpy2py(output)
    
    return output

File: backend/graphs/util/proportions.R
library(dplyr)

proportions = function(data, group_list, word_list, group, word, n) {
  word = dplyr::enquo(word)
  group = dplyr::enquo(group)
  n = dplyr::enquo(n)
  
  # Get the total word counts for each group
  group_counts = data %>%
    dplyr::group_by(group) %>%
    dplyr::summarise(n = sum(n))
  
  # Create a subset of data with the words in word_list
  # If word_list is empty, use all words
  if (length(word_list) > 0) {
    data_subset = data %>%
        dplyr::filter(word %in% word_list)
  } else {
    data_subset = data
  }
  
  # If data_subset is empty, end the function
  if (nrow(data_subset) == 0) {
    return(data.frame())
  }
  
  # Calculate word use proportion for each word/group combination
  proportions = c()
  for (i in 1:nrow(data_subset)) {
    # Get the total word count of the current group
    curr_count = group_counts %>%
      dplyr::filter(group == data_subset$group[i]) %>%
      head(n = 1) %>%
      dplyr::pull(n)
    
    # Calculate word use proportion
    value = data_subset$n[i] / curr_count
    proportions = c(proportions, value)
  }
  
  # Add proportions column
  data_subset = data_subset %>%
    dplyr::mutate(proportion = proportions) %>%
    dplyr::select(-id, -n)
  
  return(data_subset)
}

proportions_all_groups = function(data, word_list, word, n) {
  word = dplyr::enquo(word)
  n = dplyr::enquo(n)
  
  # Get the total word count
  total_count = data %>%
    dplyr::summarise(n = sum(n)) %>%
    dplyr::pull(n)
  
  # Create a subset of data with the words in word_list
  # If word_list is empty, use all words
  if (length(word_list) > 0) {
    data_subset = data %>%
        dplyr::filter(word %in% word_list)
  } else {
    data_subset = data
  }
  
  # If data_subset is empty, end the function
  if (nrow(data_subset) == 0) {
    return(data.frame())
  }
  
  # Calculate word use proportion for each word/group combination
  proportions = c()
  for (i in 1:nrow(data_subset)) {    
    # Calculate word use proportion
    value = data_subset$n[i] / total_count
    proportions = c(proportions, value)
  }
  
  # Add proportions column
  data_subset = data_subset %>%
    dplyr::mutate(proportion = proportions) %>%
    dplyr::select(-id, -n)
  
  return(data_subset)
}